common/raw_hashtable_metadata_group.h

```diff
@@ -0,0 +1,930 @@
+    for (ssize_t i : llvm::seq<ssize_t>(0, Size)) {
+      if (bytes[i] == Empty) {
+        return MatchIndex(bit);
+      }
+      bit <<= 1;
+    }
+    return MatchIndex(0);
+  }
+
+  // Materialize the group into a word.
+  uint64_t mask = byte_ints[0] | (byte_ints[0] << 7);
+  mask = ~mask & MSBs;
+
+  CARBON_DCHECK(
+      VerifyIndexMask(mask, [](uint8_t byte) { return byte == Empty; }));
+  return MatchIndex(mask);
+}
+
+inline auto MetadataGroup::PortableMatchDeleted() const -> MatchIndex {
+  if constexpr (Size > 8) {
+    static_assert(Size <= 32, "Sizes larger than 32 not yet supported!");
+    uint32_t bit = 1;
+    for (ssize_t i : llvm::seq<ssize_t>(0, Size)) {
+      if (bytes[i] == Deleted) {
+        return MatchIndex(bit);
+      }
+      bit <<= 1;
+    }
+    return MatchIndex(0);
+  }
+
+  // Materialize the group into a word.
+  uint64_t mask = byte_ints[0] | (~byte_ints[0] << 7);
+  mask = ~mask & MSBs;
+
+  CARBON_DCHECK(
+      VerifyIndexMask(mask, [](uint8_t byte) { return byte == Deleted; }));
+  return MatchIndex(mask);
+}
+
+inline auto MetadataGroup::PortableCompareEqual(MetadataGroup lhs,
+                                                MetadataGroup rhs) -> bool {
+  return llvm::equal(lhs.bytes, rhs.bytes);
+}
+
+inline auto MetadataGroup::SIMDLoad(const uint8_t* metadata, ssize_t index)
+    -> MetadataGroup {
+  MetadataGroup g;
+#if CARBON_NEON_SIMD_SUPPORT
+  g.byte_vec = vld1_u8(metadata + index);
+#elif CARBON_X86_SIMD_SUPPORT
+  g.byte_vec =
+      _mm_load_si128(reinterpret_cast<const __m128i*>(metadata + index));
+#else
+  static_assert(!UseSIMD, "Unimplemented SIMD operation");
+  static_cast<void>(metadata);
+  static_cast<void>(index);
+#endif
+  return g;
+}
+
+inline auto MetadataGroup::SIMDStore(uint8_t* metadata, ssize_t index) const
+    -> void {
+#if CARBON_NEON_SIMD_SUPPORT
+  vst1_u8(metadata + index, byte_vec);
+#elif CARBON_X86_SIMD_SUPPORT
+  _mm_store_si128(reinterpret_cast<__m128i*>(metadata + index), byte_vec);
+#else
+  static_assert(!UseSIMD, "Unimplemented SIMD operation");
+  static_cast<void>(metadata);
+  static_cast<void>(index);
+#endif
+}
+
+inline auto MetadataGroup::SIMDClearDeleted() -> void {
+#if CARBON_NEON_SIMD_SUPPORT
+  byte_ints[0] &= (~LSBs | byte_ints[0] >> 7);
+#elif CARBON_X86_SIMD_SUPPORT
+  byte_vec = _mm_blendv_epi8(_mm_setzero_si128(), byte_vec, byte_vec);
+#else
+  static_assert(!UseSIMD && !DebugSIMD, "Unimplemented SIMD operation");
+#endif
+}
+
+inline auto MetadataGroup::SIMDMatch(uint8_t present_byte) const -> MatchRange {
+  MatchRange result;
+#if CARBON_NEON_SIMD_SUPPORT
+  auto match_byte_vec = vdup_n_u8(present_byte | PresentMask);
+  auto match_byte_cmp_vec = vceq_u8(byte_vec, match_byte_vec);
+  uint64_t mask = vreinterpret_u64_u8(match_byte_cmp_vec)[0];
+  // Mask in scalar to fold with testing for zero.
+  result = MatchRange(mask & MSBs);
+#elif CARBON_X86_SIMD_SUPPORT
+  result = X86SIMDMatch(present_byte | PresentMask);
```

Comment:
Any reason this is a separate function rather than inlined here?

Reply:
Because it has subtly different semantics from what `Match` needs. I wanted to keep `SIMDMatch` and `PortableMatch` exactly the same contract as `Match`. But for `PortableMatch`, we specifically don't want to allow `Empty` or `Deleted`. But on X86, we have a common implementation we want to use for all three cases.